cargo-ndk/src/apk.rs

```diff
@@ -0,0 +1,113 @@
+use crate::manifest::Manifest;
+use android_build_tools::apk::{Apk, ApkConfig};
+use android_build_tools::config::Config;
+use android_build_tools::error::NdkError;
+use android_build_tools::ndk::Ndk;
+use android_build_tools::target::Target;
+use cargo_subcommand::{Artifact, CrateType, Subcommand};
+use std::path::PathBuf;
+use std::process::Command;
+
+pub struct ApkBuilder<'a> {
+    cmd: &'a Subcommand,
+    ndk: Ndk,
+    manifest: Manifest,
+    build_dir: PathBuf,
+    build_targets: Vec<Target>,
+}
+
+impl<'a> ApkBuilder<'a> {
+    pub fn from_subcommand(cmd: &'a Subcommand) -> Result<Self, NdkError> {
+        let ndk = Ndk::from_env()?;
+        let manifest = Manifest::parse_from_toml(cmd.manifest())?;
+        let build_targets = if let Some(target) = cmd.target() {
+            vec![Target::from_rust_triple(target)?]
+        } else {
+            vec![ndk.detect_abi().unwrap_or(Target::Arm64V8a)]
+        };
+        let build_dir = cmd.target_dir().join(cmd.profile()).join("ndk");
+        Ok(Self {
+            cmd,
+            ndk,
+            manifest,
+            build_dir,
+            build_targets,
+        })
+    }
+
+    pub fn cargo_build(&self, ndk: &Ndk, target: Target, sdk_version: u32) -> Result<(), NdkError> {
+        let triple = target.rust_triple();
+        let mut cargo = Command::new("cargo");
+        cargo.arg("build");
+
+        let clang = ndk.clang(target, sdk_version)?;
+        cargo.env(format!("CC_{}", triple), &clang);
+        cargo.env(cargo_env_target_cfg("LINKER", triple), &clang);
+
+        let ar = ndk.toolchain_bin("ar", target)?;
+        cargo.env(format!("AR_{}", triple), &ar);
+        cargo.env(cargo_env_target_cfg("AR", triple), &ar);
+
+        if self.cmd.target().is_none() {
+            cargo.arg("--target").arg(triple);
+        }
+
+        cargo.args(self.cmd.args()).status()?;
+        Ok(())
+    }
+
+    pub fn build(&self, artifact: &Artifact) -> Result<Apk, NdkError> {
+        let config = Config {
+            ndk: self.ndk.clone(),
+            build_dir: self.build_dir.join(artifact),
+            package_name: package_name(artifact),
+            package_label: artifact.name().to_string(),
+            version_name: self.manifest.version.clone(),
```

Comment:
I think we should support setting apk version using `version_code`/`version_name` (from `package.metadata.android`).
In case when this fields is not set we simply generate it from cargo package metadata.

Reply:
Oh it seems this is defaults. Okay.

Reply:
The reason why the package name and version info can't be set in metadata.android is because it's unclear what artifact they refer to (how can we set the package_name of an example) and in a more complex setup you may want to create multiple apks from the same crate (instead of a fat apk for example) with multiple version codes.